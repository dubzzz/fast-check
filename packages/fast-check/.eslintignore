bundle/
node_modules/
lib/
lib-*/
perf/
jest*.js
jest*.cjs