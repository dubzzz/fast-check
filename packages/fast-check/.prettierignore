bundle/
coverage/
docs/
documentation/Arbitraries.md
lib/
lib-*/