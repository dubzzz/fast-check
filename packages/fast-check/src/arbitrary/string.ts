import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import type { ArrayConstraintsInternal } from './array';
import { array } from './array';
import { char } from './char';
import type { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
import { createSlicesForString } from './_internals/helpers/SlicesForStringBuilder';
export type { StringSharedConstraints } from './_shared/StringSharedConstraints';

const safeObjectAssign = Object.assign;

/**
 * For strings of {@link char}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
export function string(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  const charArbitrary = char();
  const experimentalCustomSlices = createSlicesForString(charArbitrary, codePointsToStringUnmapper);
  // TODO - Move back to object spreading as soon as we bump support from es2017 to es2018+
  const enrichedConstraints: ArrayConstraintsInternal<string> = safeObjectAssign(safeObjectAssign({}, constraints), {
    experimentalCustomSlices,
  });
  return array(charArbitrary, enrichedConstraints).map(codePointsToStringMapper, codePointsToStringUnmapper);
}
