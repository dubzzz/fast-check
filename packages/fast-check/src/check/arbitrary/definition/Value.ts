import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
