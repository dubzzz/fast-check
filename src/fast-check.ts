import * as fc from './fast-check-default';
export default fc;

export * from './fast-check-default';
