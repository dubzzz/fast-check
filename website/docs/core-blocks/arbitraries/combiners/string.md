---
slug: /core-blocks/arbitraries/combiners/string/
---

# String

Combine and enhance existing string arbitraries.

## stringOf

String containing characters produced by the passed character generator.

**Signatures:**

- `fc.stringOf(charArb)`
- `fc.stringOf(charArb, {minLength?, maxLength?, size?})`

**with:**

- `charArb` — _arbitrary able to generate random strings (possibly multiple characters)_
- `minLength?` — default: `0` — _minimal number of characters (included)_
- `maxLength?` — default: `0x7fffffff` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _maximal number of characters (included)_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

**Usages:**

```js
fc.stringOf(fc.hexa());
// Examples of generated values: "f312", "ab286", "6e1a5", "5e3", "9"…

fc.stringOf(fc.char(), { maxLength: 3 });
// Note: Any string containing up to 3 (included) characters extracted from `fc.char()`
// Examples of generated values: "", "W|", "J&", "x#", "\""…

fc.stringOf(fc.char(), { minLength: 4, maxLength: 6 });
// Note: Any string containing between 4 (included) and 6 (included) characters extracted from `fc.char()`
// Examples of generated values: "j1,p{h", "[~%?", "&alf", "call!", "\"&S \"!"…

fc.stringOf(fc.constantFrom('a', 'b'), { maxLength: 5 });
// Note: Any string containing between 0 (included) and 5 (included) characters extracted from `fc.constantFrom('a', 'b')`
// Examples of generated values: "ba", "bb", "aba", "", "abb"…

fc.stringOf(fc.constantFrom('Hello', 'World'), { minLength: 1, maxLength: 3 });
// Note: It might produce strings like "WorldWorldWorld" or "WorldHelloWorld"…
// Examples of generated values: "Hello", "World", "HelloWorld", "WorldWorldHello", "HelloWorldHello"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/stringOf.html).  
Available since 1.1.3.

## stringMatching

String matching the passed regex.

**Signatures:**

- `fc.stringMatching(regex)`
- `fc.stringMatching(regex, {size?})`

**with:**

- `regex` — _the regex to be matched by produced strings_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

**Usages:**

```js
fc.stringMatching(/\s(html|php|css|java(script)?)\s/);
// Note: The regex does not contain ^ or $ assertions, so extra text could be added before and after the match
// Examples of generated values: "ca\rjava U4", "'Kn7&cP<5:\tjava\n", "<NfX\rcss\r*.", "%\u000bjavascript\fname", "#\u000bcss\u000b`&HpS"…

fc.stringMatching(/^rgb\((?:\d|[1-9]\d|1\d\d|2[0-5]\d),(?:\d|[1-9]\d|1\d\d|2[0-5]\d),(?:\d|[1-9]\d|1\d\d|2[0-5]\d)\)$/);
// Note: Regex matching RGB colors
// Examples of generated values: "rgb(237,6,11)", "rgb(143,160,247)", "rgb(257,213,251)", "rgb(4,185,33)", "rgb(253,230,211)"…

fc.stringMatching(/^[0-9a-f]{8}-[0-9a-f]{4}-[12345][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/);
// Note: Regex matching UUID
// Examples of generated values:
// • "fd606aa1-b53b-1c7b-9e2f-1e2c1ff1b8e9"
// • "e74cec0b-bd5a-4dba-96a9-edbfa9c1a198"
// • "fcccdcf3-908e-5179-adce-7ebae72c12dc"
// • "0eab1fab-5bc2-336c-9ccb-a3fecbe72ee2"
// • "bb3073ee-2283-2538-ba0c-1b976ebb9610"
// • …

fc.stringMatching(
  /^(?:\d|[1-9]\d|1\d\d|2[0-5]\d)\.(?:\d|[1-9]\d|1\d\d|2[0-5]\d)\.(?:\d|[1-9]\d|1\d\d|2[0-5]\d)\.(?:\d|[1-9]\d|1\d\d|2[0-5]\d)$/,
);
// Note: Regex matching IP v4, we rather recommend you to rely on `fc.ipV4()`
// Examples of generated values: "226.4.220.240", "206.2.148.227", "247.32.128.41", "165.252.212.135", "18.225.51.96"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/stringMatching.html).  
Available since 3.10.0.

## mixedCase

Switch the case of characters generated by an underlying arbitrary.

**Signatures:**

- `fc.mixedCase(stringArb)`
- `fc.mixedCase(stringArb, { toggleCase?, untoggleAll? })`

**with:**

- `stringArb` — _arbitrary producing random strings_
- `toggleCase?` — default: _try `toUpperCase` on the received code-point, if no effect try `toLowerCase`_ — _custom toggle case function that will be called on some of the code-points to toggle the character_
- `untoggleAll?` — default: `undefined` — _transform a string containing possibly toggled items to its untoggled version, when provided it makes it possible to shrink user-definable values, otherwise user-definable values will not be shrinkable BUT values generated by the framework will be shrinkable_

**Usages:**

```js
fc.mixedCase(fc.hexaString());
// Examples of generated values: "", "C63", "064", "1e412E00", "0E"…

fc.mixedCase(fc.constant('hello world'));
// Examples of generated values: "HEllO wOrLd", "hElLo WoRLD", "hELlo woRlD", "helLO WOrLd", "HEllo wOrld"…

fc.mixedCase(fc.constant('hello world'), {
  toggleCase: (rawChar) => `UP(${rawChar})`,
  // untoggleAll is optional, we use it in this example to show how to use all the options together
  untoggleAll: (toggledString) => toggleString.replace(/UP\((.)\)/g, '$1'),
});
// Examples of generated values:
// • "UP(h)elUP(l)o UP(w)UP(o)rUP(l)UP(d)"
// • "UP(h)eUP(l)UP(l)UP(o) UP(w)oUP(r)UP(l)UP(d)"
// • "UP(h)UP(e)lUP(l)UP(o)UP( )UP(w)UP(o)UP(r)ld"
// • "UP(h)elUP(l)oUP( )UP(w)orUP(l)UP(d)"
// • "helUP(l)o UP(w)orlUP(d)"
// • …

fc.mixedCase(fc.constant('🐱🐢🐱🐢🐱🐢'), {
  toggleCase: (rawChar) => (rawChar === '🐱' ? '🐯' : '🐇'),
});
// Examples of generated values: "🐯🐇🐱🐢🐯🐢", "🐱🐇🐱🐇🐯🐇", "🐱🐢🐯🐢🐱🐢", "🐱🐢🐱🐇🐯🐢", "🐱🐢🐯🐢🐱🐇"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/mixedCase.html).  
Available since 1.17.0.
