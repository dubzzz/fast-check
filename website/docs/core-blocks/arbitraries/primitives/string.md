---
slug: /core-blocks/arbitraries/primitives/string/
---

# String

Generate string values.

:::tip
If you want to join several strings together: refer to our [combiners section](/docs/core-blocks/arbitraries/combiners/). We have some [built-in combiners working exclusively on string values](/docs/core-blocks/arbitraries/combiners/string/).
:::

## hexaString

Hexadecimal string containing characters produced by `fc.hexa()`.

**Signatures:**

- `fc.hexaString()`
- `fc.hexaString({minLength?, maxLength?, size?})`

**with:**

- `minLength?` — default: `0` — _minimal number of characters (included)_
- `maxLength?` — default: `0x7fffffff` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _maximal number of characters (included)_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

**Usages:**

```js
fc.hexaString();
// Examples of generated values: "251971", "", "a9", "742e6c86e", "39350b163"…

fc.hexaString({ maxLength: 3 });
// Note: Any hexadecimal string containing up to 3 (included) characters
// Examples of generated values: "1", "", "2ef", "2a", "6e3"…

fc.hexaString({ minLength: 3 });
// Note: Any hexadecimal string containing at least 3 (included) characters
// Examples of generated values: "1021a292c2d306", "e4660fd014ae290", "2ef914a5d7ffe9df", "2a212", "05dd1"…

fc.hexaString({ minLength: 4, maxLength: 6 });
// Note: Any hexadecimal string containing between 4 (included) and 6 (included) characters
// Examples of generated values: "b4ccb", "e51d", "b3e093", "383f", "27bd"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/hexaString.html).  
Available since 0.0.1.

## string

String containing characters produced by `fc.char()`.

**Signatures:**

- `fc.string()`
- `fc.string({minLength?, maxLength?, size?})`

**with:**

- `minLength?` — default: `0` — _minimal number of characters (included)_
- `maxLength?` — default: `0x7fffffff` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _maximal number of characters (included)_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

**Usages:**

```js
fc.string();
// Examples of generated values: "JT>\"C9k", "h]iD\"27;", "S", "n\\Ye", ""…

fc.string({ maxLength: 3 });
// Note: Any string containing up to 3 (included) characters
// Examples of generated values: "", "ref", "?D", "key", "}"…

fc.string({ minLength: 3 });
// Note: Any string containing at least 3 (included) characters
// Examples of generated values: "Pv-^X_t", "bind", "?DM", "iEjK.b?^O", "}~}S"…

fc.string({ minLength: 4, maxLength: 6 });
// Note: Any string containing between 4 (included) and 6 (included) characters
// Examples of generated values: "Trxall", "&&@%4", "s@IO", "0\"zM", "}#\"$"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/string.html).  
Available since 0.0.1.

## asciiString

ASCII string containing characters produced by `fc.ascii()`.

**Signatures:**

- `fc.asciiString()`
- `fc.asciiString({minLength?, maxLength?, size?})`

**with:**

- `minLength?` — default: `0` — _minimal number of characters (included)_
- `maxLength?` — default: `0x7fffffff` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _maximal number of characters (included)_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

**Usages:**

```js
fc.asciiString();
// Examples of generated values: "\f@D", "hp", "q#dO~?@", "Qad", "5eHqc"…

fc.asciiString({ maxLength: 3 });
// Note: Any ascii string containing up to 3 (included) characters
// Examples of generated values: "6", "", "ty", ",", "k"…

fc.asciiString({ minLength: 3 });
// Note: Any ascii string containing at least 3 (included) characters
// Examples of generated values: "603e", "6W\u001b^tR-\n\n|", "efproto_\u001abhasOw", "$\u001c&\u0000R", "apply"…

fc.asciiString({ minLength: 4, maxLength: 6 });
// Note: Any ascii string containing between 4 (included) and 6 (included) characters
// Examples of generated values: "<&\u001e\u001b ", "bind", "dnGn\\2", "& % !", "__defi"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/asciiString.html).  
Available since 0.0.1.

## unicodeString

Unicode string containing characters produced by `fc.unicode()`.

**Signatures:**

- `fc.unicodeString()`
- `fc.unicodeString({minLength?, maxLength?, size?})`

**with:**

- `minLength?` — default: `0` — _minimal number of characters (included)_
- `maxLength?` — default: `0x7fffffff` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _maximal number of characters (included)_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

**Usages:**

```js
fc.unicodeString();
// Examples of generated values: ",", "뇇ⷍ돠", "", "穿ﺥ羊汉ힸ౨", "穙셣킲햗䊢鏶눏"…

fc.unicodeString({ maxLength: 3 });
// Note: Any unicode (from BMP-plan) string containing up to 3 (included) characters
// Examples of generated values: "", "꙰", "툌", "㮟୆", "a"…

fc.unicodeString({ minLength: 3 });
// Note: Any unicode (from BMP-plan) string containing at least 3 (included) characters
// Examples of generated values: "떚碱縢鰷釬ڤ撙", "竑૱toS", "꙰ꁺ蜱⫄ั", "__d", "툌뀪凛瞸㾡끴"…

fc.unicodeString({ minLength: 4, maxLength: 6 });
// Note: Any unicode (from BMP-plan) string containing between 4 (included) and 6 (included) characters
// Examples of generated values: "뿐噻⪃嫿垈", "ڣۡ觌뱇", "apply", "￹톥薦￾", "namea"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/unicodeString.html).  
Available since 0.0.11.

## string16bits

String containing characters produced by `fc.char16bits()`.

Be aware that the generated string might appear invalid regarding the unicode standard as it might contain incomplete pairs of surrogate.

**Signatures:**

- `fc.string16bits()`
- `fc.string16bits({minLength?, maxLength?, size?})`

**with:**

- `minLength?` — default: `0` — _minimal number of characters (included)_
- `maxLength?` — default: `0x7fffffff` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _maximal number of characters (included)_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

**Usages:**

```js
fc.string16bits();
// Examples of generated values: "", "沉蹻!쯼&)￹噕￶￻", "獏", "嚷爇꡽邏䨫Ꝺ䟌", "ۆ딛楯씺"…

fc.string16bits({ maxLength: 3 });
// Note: Any string (not really legal ones sometimes) containing up to 3 (included) characters
// Examples of generated values: "", "ꃷ", "톽va", "뿤䵎悧", ""…

fc.string16bits({ minLength: 3 });
// Note: Any string (not really legal ones sometimes) containing at least 3 (included) characters
// Examples of generated values: "㝟佷㟧࿝譽먔", "ꃷ￱￷￻ꢒ￺￸", "톽valueOf", "key", "app"…

fc.string16bits({ minLength: 4, maxLength: 6 });
// Note: Any string (not really legal ones sometimes) containing between 4 (included) and 6 (included) characters
// Examples of generated values: "apply", "鹽\udc68鯻שּׂ", "땺\uda2f熑鉈뗻", "__def", "㓐줫曧ᒢ"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/string16bits.html).  
Available since 0.0.11.

## fullUnicodeString

Unicode string containing characters produced by `fc.fullUnicode()`.

**Signatures:**

- `fc.fullUnicodeString()`
- `fc.fullUnicodeString({minLength?, maxLength?, size?})`

**with:**

- `minLength?` — default: `0` — _minimal number of characters (included)_
- `maxLength?` — default: `0x7fffffff` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _maximal number of characters (included)_
- `size?` — default: `undefined` [more](/docs/configuration/larger-entries-by-default/#size-explained) — _how large should the generated values be?_

_Be aware that the length is considered in terms of the number of glyphs in the string and not the number of UTF-16 characters. As a consequence `generatedString.length` might be greater than the asked maximal length but `[...generatedString].length` will not and always be in the required range_

**Usages:**

```js
fc.fullUnicodeString();
// Examples of generated values: "򎅯񳓃󲢱򟭂", "󁵺涫򣤲󋣈󪷆", "󋅶񏵞󒽡󫺊񏽵", "􍄵򒂾󃉓", "񡚈"…

fc.fullUnicodeString({ maxLength: 3 });
// Note: Any unicode string containing up to 3 (included) code-points
// Examples of generated values: "򅉞񸟤󐋚", "􏿴", "􇢺", "", "󏫭"…

fc.fullUnicodeString({ minLength: 3 });
// Note: Any unicode string containing at least 3 (included) code-points
// Examples of generated values: "򅉞񸟤󐋚򲮃󺮞􍶈󂾓񅙎򼝄󒂁񌸩򌻜󎼿񺡓", "bind", "􇢺󣓺in", "%4􏿻propertyIs", "𒝰􇋧􏿫􏿱񧌫"…

fc.fullUnicodeString({ minLength: 4, maxLength: 6 });
// Note: Any unicode string containing between 4 (included) and 6 (included) code-points
// Examples of generated values: "call", "񛰖񞑑󱈋𨤎", "񉓁򔶍򣵵乀򠍾󢏘", ",valu", "󐣙󼃞񫎢󖫩𫅰𪂀"…
```

Resources: [API reference](https://fast-check.dev/api-reference/functions/fullUnicodeString.html).  
Available since 0.0.11.
