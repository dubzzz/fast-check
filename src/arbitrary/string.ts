import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { convertFromNext, convertToNext } from '../check/arbitrary/definition/Converters';
import { array } from './array';
import { char } from './char';
import { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
export { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For strings of {@link char}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
export function string(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  return convertFromNext(
    convertToNext(array(char(), constraints)).map(codePointsToStringMapper, codePointsToStringUnmapper)
  );
}
