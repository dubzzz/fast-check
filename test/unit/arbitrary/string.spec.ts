import * as fc from '../../../lib/fast-check';
import { string } from '../../../src/arbitrary/string';

import { Value } from '../../../src/check/arbitrary/definition/Value';
import {
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
} from './__test-helpers__/ArbitraryAssertions';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';

describe('string (integration)', () => {
  type Extra = { minLength?: number; maxLength?: number };
  const extraParameters: fc.Arbitrary<Extra> = fc
    .tuple(fc.nat({ max: 5 }), fc.nat({ max: 30 }), fc.boolean(), fc.boolean())
    .map(([min, gap, withMin, withMax]) => ({
      minLength: withMin ? min : undefined,
      maxLength: withMax ? min + gap : undefined,
    }));

  const isCorrect = (value: string, extra: Extra) => {
    if (extra.minLength !== undefined) {
      expect(value.length).toBeGreaterThanOrEqual(extra.minLength);
    }
    if (extra.maxLength !== undefined) {
      expect(value.length).toBeLessThanOrEqual(extra.maxLength);
    }
    for (const c of value.split('')) {
      expect(c.charCodeAt(0)).toBeGreaterThanOrEqual(0x20);
      expect(c.charCodeAt(0)).toBeLessThanOrEqual(0x7e);
    }
  };

  const stringBuilder = (extra: Extra) => string(extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(stringBuilder, { extraParameters });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(stringBuilder, isCorrect, { extraParameters });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(stringBuilder, { extraParameters });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(stringBuilder, { extraParameters });
  });

  it.each`
    source
    ${''}
    ${'azerty'}
    ${'ah! ah!'}
    ${'0123456789' /* by default maxLength = maxLengthFromMinLength(0) = 10 */}
  `('should be able to generate $source with fc.string()', ({ source }) => {
    // Arrange / Act
    const arb = string();
    const out = arb.canShrinkWithoutContext(source);

    // Assert
    expect(out).toBe(true);
  });

  it.each`
    source                                        | constraints
    ${'a\u{1f431}b' /* out-of-range character */} | ${{}}
    ${'ab' /* not large enough */}                | ${{ minLength: 3 }}
    ${'abcd' /* too large */}                     | ${{ maxLength: 3 }}
  `('should not be able to generate $source with fc.string($constraints)', ({ source, constraints }) => {
    // Arrange / Act
    const arb = string(constraints);
    const out = arb.canShrinkWithoutContext(source);

    // Assert
    expect(out).toBe(false);
  });

  it.each`
    rawValue
    ${'Hey!'}
    ${'0'.repeat(50) /* longer than default maxGeneratedLength but ok for shrink */}
  `('should be able to shrink $rawValue with fc.string()', ({ rawValue }) => {
    // Arrange
    const arb = string();
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
