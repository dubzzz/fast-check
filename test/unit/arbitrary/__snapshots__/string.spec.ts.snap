// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string (integration) should be able to shrink Hey! 1`] = `
"\\"Hey!\\"
├> \\"\\"
├> \\"y!\\"
|  ├> \\"!\\"
|  |  ├> \\"\\"
|  |  └> \\" \\"
|  |     └> \\"\\"
|  ├> \\" !\\"
|  |  ├> \\"!\\"
|  |  |  ├> \\"\\"
|  |  |  └> \\" \\"
|  |  |     └> \\"\\"
|  |  ├> \\" \\"
|  |  |  └> \\"\\"
|  |  └> \\"  \\"
|  |     ├> \\" \\"
|  |     |  └> \\"\\"
|  |     └> \\" \\"
|  |        └> \\"\\"
|  ├> \\"M!\\"
|  |  ├> \\"!\\"
|  |  |  ├> \\"\\"
|  |  |  └> \\" \\"
|  |  |     └> \\"\\"
|  |  ├> \\"7!\\"
|  |  |  ├> \\"!\\"
|  |  |  |  ├> \\"\\"
|  |  |  |  └> \\" \\"
|  |  |  |     └> \\"\\"
|  |  |  ├> \\",!\\"
|  |  |  |  ├> \\"!\\"
|  |  |  |  |  ├> \\"\\"
|  |  |  |  |  └> \\" \\"
|  |  |  |  |     └> \\"\\"
|  |  |  |  ├> \\"&!\\"
|  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  ├> \\"#!\\"
|  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  ├> \\"\\\\\\"!\\"
|  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  ├> \\"!!\\"
|  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  ├> \\" !\\"
|  |  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  |  ├> \\" \\"
|  |  |  |  |  |  |  |  |  |  └> \\"\\"
|  |  |  |  |  |  |  |  |  └> \\"  \\"
|  |  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |  |     └> \\" \\"
|  |  |  |  |  |  |  |  |        └> \\"\\"
|  |  |  |  |  |  |  |  ├> \\"!\\"
|  |  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  |  └> \\" \\"
|  |  |  |  |  |  |  |  |     └> \\"\\"
|  |  |  |  |  |  |  |  └> \\"! \\"
|  |  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |  |     ├> \\"  \\"
|  |  |  |  |  |  |  |     |  ├> \\" \\"
|  |  |  |  |  |  |  |     |  |  └> \\"\\"
|  |  |  |  |  |  |  |     |  └> \\" \\"
|  |  |  |  |  |  |  |     |     └> \\"\\"
|  |  |  |  |  |  |  |     └> \\"!\\"
|  |  |  |  |  |  |  |        ├> \\"\\"
|  |  |  |  |  |  |  |        └> \\" \\"
|  |  |  |  |  |  |  |           └> \\"\\"
|  |  |  |  |  |  |  ├> \\"\\\\\\"\\"
|  |  |  |  |  |  |  |  ├> \\"\\"
|  |  |  |  |  |  |  |  └> \\"!\\"
|  |  |  |  |  |  |  |     ├> \\"\\"
|  |  |  |  |  |  |  |     └> \\" \\"
|  |  |  |  |  |  |  |        └> \\"\\"
|  |  |  |  |  |  |  └> \\"\\\\\\" \\"
|  |  |  |  |  |  |     ├> \\" \\"
|  |  |  |  |  |  |     |  └> \\"\\"
|  |  |  |  |  |  |     ├> \\"! \\"
|  |  |  |  |  |  |     |  ├> \\" \\"
|  |  |  |  |  |  |     |  |  └> \\"\\"
|  |  |  |  |  |  |     |  ├> \\"  \\"
|  |  |  |  |  |  |     |  |  ├> \\" \\"
|  |  |  |  |  |  |     |  |  |  └> \\"\\"
|  |  |  |  |  |  |     |  |  └> \\" \\"
|  |  |  |  |  |  |     |  |     └> \\"\\"
|  |  |  |  |  |  |     |  └> \\"!\\"
|  |  |  |  |  |  |     |     └> …
|  |  |  |  |  |  |     └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
